### ISSUE TITLE
Buy value of StochRSI, RSI, Emagain, Emacross become ZERO permanently. Have to restart to fix.

### ISSUE THREAD

**Issue by leminhduc** [taniman/profit-trailer#2108]
NO NEW FEATURES! Use [PT Enahancements](https://github.com/taniman/profit-trailer-enhancements) for Feature Requests.

## ProfitTrailer info
Version: 2.1.20 and also happen in the previous version. I upgraded to the latest version to see if it fixed, but the problem still occur

 
Exchange:
- [ ] Bittrex
- [x] Binance
- [ ] Poloniex

 
Market:
- [x] BTC
- [ ] ETH  (ETH base works fine)
- [ ] USDT
- [ ] BNB
- [ ] XMR

 
Operating System:
- [x] Windows (please specify version)
- [ ] Ubuntu (please specify version)
- [ ] Other linux distro (please specify)
- [ ] Mac OS

## Expected Behavior

Emagain (ShortGain), RSI, Stochrsi, EmaCross (ShortTermTrend) should have values changing. 

This is the screenshot of normal behavior in both Possible Buy and DCA log
![image](https://user-images.githubusercontent.com/278861/47889248-f9555000-de83-11e8-8d9c-3509f8eb1fd4.png)

## Current Behavior

On BTC base, I just let the bot run for 1 or 2 days, the problem will appear again:
 All the Pairs in PossibleBuy and DCA log have Emagain (ShortGain), RSI, Stochrsi, EmaCross (ShortTermTrend) value ZERO. This value will not change until I restart the bot. 
This is a serious issue that blocks the bot from buying first pair or DCA 
![image](https://user-images.githubusercontent.com/278861/47889174-88ae3380-de83-11e8-815f-5e9f0a133b6b.png)

## Troubleshooting

Restart the bot, all the values work again. But after sometimes (1 day) it will all go back to ZERO

## Steps to Reproduce

1.
2.
3.
4.

## Configuration

It is not the issue of Configuration, because I use the same settings for both ETH and BTC. This setting for BTC has been working fine, until recently. 

## Log files/console output

 

[log_2108.zip](https://github.com/taniman/profit-trailer/files/2540774/log_2108.zip)

**Comment by leminhduc** [comment]
Screenshot comparision of ETH bot and BTC bot 
![image](https://user-images.githubusercontent.com/278861/47890203-ccf00280-de88-11e8-9606-d0ad14d1761c.png)

**Comment by leminhduc** [comment]
This is the screenshot of 2 BTC bots in the same server, the left has error, and the right is fine 
![image](https://user-images.githubusercontent.com/278861/47890527-345a8200-de8a-11e8-9c3e-1715362ed836.png)

**Comment by taniman** [comment]
Please provide your settings I will see If i notice something while running same settings.

**Comment by nealgentry** [comment]
I recently had the same issue with one of my bots. I was able to fix by installing a fresh copy of ProfitTrailer... However, you will lose your history.

**Comment by leminhduc** [comment]
oh wow, maybe I have to do that, this bot has become faulty. Yesterday, it keeps buying in 2 new bags even though the max pairs already exceeded. I have to restart it and it stops buying.

**Comment by leminhduc** [comment]
Yes, I had lots of memory allocation errors with this bot too. Not sure why memory issues could affect the logic hmm

**Comment by leminhduc** [comment]
The entire 140 pairs in the Possible list and DCA list are affected, not just one or two coins. I suspect lack of memory/CPU kills the process/thread for all the Short timeframe indicators (5-min), as the longer timeframe (2h, 4h, 1D work fine)

**Comment by nealgentry** [comment]
Ok, I have been experiencing this issue for a little while now. I was able to re-create it, I have 2 API keys in use on the same binance account. I stopped live trading due to being unable to solve this issue.

1. I downloaded a complete fresh 2.1.21 unused copy of PT, I was able to still recreate the issue by using my Original settings and Original API key.  

2. How I was able to avoid the problem; I created a new settings slot, I copied the old settings to a new name, i created a new API key, I loaded PT version 2.1.11 under a new pm2 name. 4 days up NO PROBLEMS....

**Comment by leminhduc** [comment]
Thank a lot bro, I will try your approach.

**Comment by nealgentry** [comment]
As you can see NEBLBTC on 2.1.11 with the new settings name and API works fine.
NEBLBTC on 2.1.21 with a fresh download locks up and has been for 1 hour so far.
![1](https://user-images.githubusercontent.com/34993553/47975644-49930480-e074-11e8-99fa-6572e36bfb77.jpg)
![2](https://user-images.githubusercontent.com/34993553/47975844-4fd5b080-e075-11e8-882b-c1514daddb65.jpg)

**Comment by leminhduc** [comment]
Holy smoky, you found the root cause!

**Comment by nealgentry** [comment]
The main reason I've stopped live trading is due to the bot locking up into a true state, purchasing multiple DCA's and needlessly draining my balance ultimately losing lots of money.

**Comment by leminhduc** [comment]
yup, it ignored my max pairs setting and keep buying 3-4 pairs until I rush home and shut it down lol

**Comment by SgtBatten** [comment]
I notice you have 5 times as many coins in your PBL for the failing version. This is memory usage increasing still it would seem

**Comment by nealgentry** [comment]
Here is an example of this happening.
![locked true](https://user-images.githubusercontent.com/34993553/47976161-c4f5b580-e076-11e8-9aec-da35a255dc3b.jpg)

**Comment by leminhduc** [comment]
definitely memory / CPU problem then, I have 20 bags, and possible buy log 140 pairs
also, I use 7 indicators in multiple timeframes. 

Your settings and bags are even much more memory/CPU intensive than mine

**Comment by nealgentry** [comment]
by changing the bot 2.1.21 to my re-created identical settings my bot instantly began to start working.
![same settings different ver works](https://user-images.githubusercontent.com/34993553/47977232-85ca6300-e07c-11e8-9aa6-638a35bf47a0.jpg)

**Comment by nealgentry** [comment]
Upon switching back to the original save slot the bot immediately locked up.
![only settings slot neal1 locks up the bot](https://user-images.githubusercontent.com/34993553/47977347-29b40e80-e07d-11e8-926d-11a856af011b.jpg)

**Comment by nealgentry** [comment]
Remember... Both settings are identical NEAL2 is just a copy of Neal1.

Running version 2.1.11 does not lock up regardless of what settings I use.

Running version 2.1.21 Locks up upon switching to Neal1 Settings.

![not locked up x2](https://user-images.githubusercontent.com/34993553/47977504-fcb42b80-e07d-11e8-9a16-6ec3e7bf7dd1.JPG)

**Comment by taniman** [comment]
I would very much like a downloaded version of NEAL2 to run in my bot to see if i get the same problem.

Cause without anyone providing settings to run against I cannot reproduce/fix this problem

**Comment by nealgentry** [comment]
Remember I am able to run any settings on  P T Ver: 2.1.11 without lockup... 

These are the settings that lock up version 2.1.16 through 2.1.21 and possibly others I've only tested 4 versions; 2.1.16, 2.1.18, 2.1.20, 2.1.21...
All with clean downloads. 

[Neal1-PAIRS(1).txt](https://github.com/taniman/profit-trailer/files/2549838/Neal1-PAIRS.1.txt)
[Neal1-INDICATORS(1).txt](https://github.com/taniman/profit-trailer/files/2549839/Neal1-INDICATORS.1.txt)
[Neal1-DCA(1).txt](https://github.com/taniman/profit-trailer/files/2549843/Neal1-DCA.1.txt)
[MACD_Cross-PAIRS.txt](https://github.com/taniman/profit-trailer/files/2549846/MACD_Cross-PAIRS.txt)
[MACD_Cross-INDICATORS.txt](https://github.com/taniman/profit-trailer/files/2549847/MACD_Cross-INDICATORS.txt)
[MACD_Cross-DCA.txt](https://github.com/taniman/profit-trailer/files/2549848/MACD_Cross-DCA.txt)

**Comment by nealgentry** [comment]
![macd locked up](https://user-images.githubusercontent.com/34993553/48015655-40934900-e0ef-11e8-9575-32a11b24fabf.jpg)

**Comment by nealgentry** [comment]
After leaving PT 2.1.11 running for 2 days with close monitoring and value documentation. I noticed one coin pair lock into a true state (CLOAKBTC) this happened right as the coin was having a "pump" forcing my bot to DCA all the dumps on the pump. All the values on (CLOAKBTC) locked to true while all other coins are organically changing values.

![locked pump](https://user-images.githubusercontent.com/34993553/48156727-93a60100-e293-11e8-915f-57b7b44e1e50.jpg)

**Comment by taniman** [comment]
@nealgentry I am running your settings right now and will see what happens. 

One thing I want to clarify is that PT does not use close price. So that arrow you drew is incorrect. 
The value of PT uses bid and ask price. 
And the value in green (left) is exactly the same as the price of PT (right). So no error there.

Another thing I want to point out is that the reason all is TRUE and no DCA yet is because BT% is -3.7% and  your buy_trigger is set to -4.8%

So in summary, there is nothing wrong in this last image. Everything is working fine for CLOAKBTC

**Comment by nealgentry** [comment]
**@taniman** Thank you for looking into this error you are welcome to log into one of my vpns once the error comes back to see for your self. 

Just to verify what was in-fact happening all was locked into a true state following the value of the coin as it fell, the major problems with this scenario; all should not have remained locked into a true state for 6 hours, once all was locked into a true state it caused multiple DCA buys that would not have taken place if the indicators were working properly, if indicators were working properly there should have been 3 DCA buys and the coin would have dropped in value without the bot making DCA buys the entire time.
 Once the coin-pair achieved a true-state from the indicators the bot should have immediately DCA purchased the coin to the desired coin price level specified in in the DCA purchase values. 
and for further verification, I launched a second bot on the same account to see if the values were changing to rule out a binance server issue.


![lockedact](https://user-images.githubusercontent.com/34993553/48242038-e28c8d00-e39e-11e8-9240-96dc71879121.jpg)

**Comment by nealgentry** [comment]
**@taniman**
I must also tell you profit trailer is very good at keeping close to my test values... so when something like this happens it's disheartening. I love the profit trailer software and see great potential for it to follow my trading style automatically.

**Comment by nealgentry** [comment]
**@taniman**
Hours later the bot still had the true values locked on CLOAK all while others organically changed.
![hours later](https://user-images.githubusercontent.com/34993553/48242293-246a0300-e3a0-11e8-9220-0f9f097992bb.jpg)

**Comment by taniman** [comment]
@nealgentry I understand and I am hoping to see the same results on my test machine in the coming hours. days so I can find out the problem.
I won't need to log into your system as just logging in does not help with debugging. :) But tnx

6 hours later cloak indeed should not still be locked I assume. But i still urge you to update to the latest versions. As I thought you had "fixed" the problem using one of the latest versions too.

When I finally figure this one out. And if there is a fix it will be on a newer version

**Comment by taniman** [comment]
@nealgentry Running a test bot with your settings 30+ online. Data still looking good here.
Contact me on discord if you want to have a loo on the server. @elroy

Maybe I am missing something

**Comment by nealgentry** [comment]
@taniman Just upgraded to the latest release and started the troubled bot. if or when the problem comes back 'usually it takes a few days' you are more than welcome to ssh into my server and check it out.  I will definitely contact you on Dis.

Thank you for your always prompt and thoughtful responses.

**Comment by leminhduc** [comment]
the bug came back to me too. I saw this error in the log

![image](https://user-images.githubusercontent.com/278861/48662003-3e00fe00-eab6-11e8-95d2-4ea0668241c2.png)



2018-11-17 22:08:16 ERROR TelegramService - Error sending your telegram message {}
java.net.SocketTimeoutException: timeout
	at okio.Okio$4.newTimeoutException(Okio.java:230) ~[okio-1.13.0.jar!/:?]
	at okio.AsyncTimeout.exit(AsyncTimeout.java:285) ~[okio-1.13.0.jar!/:?]
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:241) ~[okio-1.13.0.jar!/:?]
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:345) ~[okio-1.13.0.jar!/:?]
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:217) ~[okio-1.13.0.jar!/:?]
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:211) ~[okio-1.13.0.jar!/:?]
	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:75) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185) ~[okhttp-3.8.1.jar!/:?]
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:135) [okhttp-3.8.1.jar!/:?]
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32) [okhttp-3.8.1.jar!/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_171]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_171]
Caused by: java.net.SocketException: Socket closed
	at java.net.SocketInputStream.socketRead0(Native Method) ~[?:1.8.0_171]
	at java.net.SocketInputStream.socketRead(Unknown Source) ~[?:1.8.0_171]
	at java.net.SocketInputStream.read(Unknown Source) ~[?:1.8.0_171]
	at java.net.SocketInputStream.read(Unknown Source) ~[?:1.8.0_171]
	at sun.security.ssl.InputRecord.readFully(Unknown Source) ~[?:1.8.0_171]
	at sun.security.ssl.InputRecord.read(Unknown Source) ~[?:1.8.0_171]
	at sun.security.ssl.SSLSocketImpl.readRecord(Unknown Source) ~[?:1.8.0_171]
	at sun.security.ssl.SSLSocketImpl.readDataRecord(Unknown Source) ~[?:1.8.0_171]
	at sun.security.ssl.AppInputStream.read(Unknown Source) ~[?:1.8.0_171]
	at okio.Okio$2.read(Okio.java:139) ~[okio-1.13.0.jar!/:?]
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:237) ~[okio-1.13.0.jar!/:?]
	... 23 more

**Comment by nealgentry** [comment]
@leminhduc not to be used as a permanent fix! I am currently using crontab to auto restart the PT service every 3 hours... It has been working for 6 days with no issues.

If you're on Linux of course...

Edit cron with crontab -e and add the following:

0 */3 * * * /usr/bin/node /usr/bin/pm2 restart all

if you have more than one process like; pttracker, pt, others... you can use process id# 

0 */3 * * * /usr/bin/node /usr/local/bin/pm2 restart 0

For the schedule, use 0 */12 * * * for every 12 hours, or 0 0,12 * * * for 0:00 and 12:00 specifically. 

* 0,12 * * *, would trigger every minute of hour 0 and hour 12, 0:00, 0:01, 0:02...)

For the command use the path to node followed by the path to pm2 and the pm2 options. Use which node and which pm2 to get the path to node and pm2.

**Comment by leminhduc** [comment]
@nealgentry thank bro. As I am using window OS, I just remote and restart it. A new version of PT released, but I don't see the fix description. Maybe not yet

**Comment by taniman** [comment]
We have better ways of restarting checkign and restarting the websockets in version 2.2. 
So this error might have been fixed i nthe latest versions of 2.2
